test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/UseOtlpExporterExtensionTests.cs

```diff
@@ -0,0 +1,344 @@
+
+            builder.ConfigureTracingExporterOptions(
+                exporterOptions => exporterOptions.Endpoint = new Uri("http://signal_endpoint/traces/"));
+            builder.ConfigureTracingProcessorOptions(
+                processorOptions =>
+                {
+                    processorOptions.ExportProcessorType = ExportProcessorType.Simple;
+                    processorOptions.BatchExportProcessorOptions.ScheduledDelayMilliseconds = 1002;
+                });
+        }
+    }
+
+    [Theory]
+    [InlineData(null)]
+    [InlineData("testNamedOptions")]
+    public void UseOtlpExporterConfigurationTest(string? name)
+    {
+        var config = new ConfigurationBuilder()
+            .AddInMemoryCollection(new Dictionary<string, string?>
+            {
+                ["DefaultOptions:Endpoint"] = "http://default_endpoint/",
+                ["LoggingOptions:Endpoint"] = "http://signal_endpoint/logs/",
+                ["LoggingOptions:ExportProcessorType"] = "Simple",
+                ["LoggingOptions:BatchExportProcessorOptions:ScheduledDelayMilliseconds"] = "1000",
+                ["MetricsOptions:Endpoint"] = "http://signal_endpoint/metrics/",
+                ["MetricsOptions:TemporalityPreference"] = "Delta",
+                ["MetricsOptions:PeriodicExportingMetricReaderOptions:ExportIntervalMilliseconds"] = "1001",
+                ["TracingOptions:Endpoint"] = "http://signal_endpoint/traces/",
+                ["TracingOptions:ExportProcessorType"] = "Simple",
+                ["TracingOptions:BatchExportProcessorOptions:ScheduledDelayMilliseconds"] = "1002",
+            })
+            .Build();
+
+        var services = new ServiceCollection();
+
+        if (!string.IsNullOrEmpty(name))
+        {
+            services.AddOpenTelemetry()
+                .UseOtlpExporter(name: name, configuration: config);
+        }
+        else
+        {
+            services.AddOpenTelemetry()
+                .UseOtlpExporter(config);
+
+            name = "otlp";
+        }
+
+        using var sp = services.BuildServiceProvider();
+
+        VerifyOptionsApplied(sp, name!);
+    }
+
+    [Fact]
+    public void UseOtlpExporterSingleCallsTest()
+    {
+        var services = new ServiceCollection();
+
+        services.AddOpenTelemetry()
+            .UseOtlpExporter();
+
+        using var sp = services.BuildServiceProvider();
+
+        Assert.NotNull(sp.GetRequiredService<LoggerProvider>());
+        Assert.NotNull(sp.GetRequiredService<MeterProvider>());
+        Assert.NotNull(sp.GetRequiredService<TracerProvider>());
+    }
+
+    [Fact]
+    public void UseOtlpExporterMultipleCallsTest()
+    {
+        var services = new ServiceCollection();
+
+        services.AddOpenTelemetry()
+            .UseOtlpExporter()
+            .UseOtlpExporter();
+
+        using var sp = services.BuildServiceProvider();
+
+        Assert.Throws<NotSupportedException>(() => sp.GetRequiredService<LoggerProvider>());
```

Comment:
Does this test that the user would actually get an exception when trying to run the application? It's probably some `ServiceProvider` related deferred execution thing but I was hoping for the code to throw an exception while executing the second   call to `UseOtlpExporter`. Maybe inspect the `ServiceCollection` itself before adding `UseOtlpExporterRegistration`?

```csharp
if (services.Any((service) => service.ServiceType == typeof(UseOtlpExporterRegistration)))
{
    throw new NotSupportedException("Multiple calls to UseOtlpExporter on the same IServiceCollection are not supported.");
}

services.AddSingleton<UseOtlpExporterRegistration>();
```

Reply:
Slight problem with doing it this way.

It will work for this case:

```csharp
        services.AddOpenTelemetry()
           .UseOtlpExporter()
           .UseOtlpExporter();
```

But it won't work for this case:

```csharp
        services.AddOpenTelemetry()
           .WithTracing(builder => builder.AddOtlpExporter()) // `UseOtlpExporterRegistration` won't exist in service collection until the next line runs
           .UseOtlpExporter();
```

The question is, will the current mechanism crash the app?

So what will happen is a `NotSupportedException` is thrown when creating either `LoggerProvider`, `MeterProvider`, or `TracerProvider`.

In the case of hosting, that will happen in the `IHostedService`.

I just dropped two `UseOtlpExporter` calls into the `Examples.AspNetCore` app to verify. It does throw:

<details>
<summary>Stack trace</summary>

```
Unhandled exception. System.NotSupportedException: Multiple calls to UseOtlpExporter on the same IServiceCollection are not supported.
   at System.OpenTelemetryBuilderServiceProviderExtensions.EnsureSingleUseOtlpExporterRegistration(IServiceProvider serviceProvider) in D:\Source\open-telemetry\codeblanch-opentelemetry-dotnet\src\OpenTelemetry.Exporter.OpenTelemetryProtocol\Builder\OpenTelemetryBuilderServiceProviderExtensions.cs:line 16
   at OpenTelemetry.Exporter.OtlpExporterBuilder.<RegisterOtlpExporterServices>g__GetBuilderOptionsAndValidateRegistrations|14_6(IServiceProvider sp, String name) in D:\Source\open-telemetry\codeblanch-opentelemetry-dotnet\src\OpenTelemetry.Exporter.OpenTelemetryProtocol\Builder\OtlpExporterBuilder.cs:line 248
   at OpenTelemetry.Exporter.OtlpExporterBuilder.<>c__DisplayClass14_0.<RegisterOtlpExporterServices>b__3(IServiceProvider sp, LoggerProviderBuilder logging) in D:\Source\open-telemetry\codeblanch-opentelemetry-dotnet\src\OpenTelemetry.Exporter.OpenTelemetryProtocol\Builder\OtlpExporterBuilder.cs:line 198
   at OpenTelemetry.Logs.OpenTelemetryDependencyInjectionLoggingServiceCollectionExtensions.ConfigureLoggerProviderBuilderCallbackWrapper.ConfigureBuilder(IServiceProvider serviceProvider, LoggerProviderBuilder loggerProviderBuilder) in D:\Source\open-telemetry\codeblanch-opentelemetry-dotnet\src\OpenTelemetry.Api.ProviderBuilderExtensions\Logs\OpenTelemetryDependencyInjectionLoggingServiceCollectionExtensions.cs:line 175
   at OpenTelemetry.Logs.LoggerProviderSdk..ctor(IServiceProvider serviceProvider, Boolean ownsServiceProvider) in D:\Source\open-telemetry\codeblanch-opentelemetry-dotnet\src\OpenTelemetry\Logs\LoggerProviderSdk.cs:line 50
   at OpenTelemetry.Logs.LoggerProviderBuilderBase.<>c.<.ctor>b__3_0(IServiceProvider sp) in D:\Source\open-telemetry\codeblanch-opentelemetry-dotnet\src\OpenTelemetry\Logs\Builder\LoggerProviderBuilderBase.cs:line 42
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite callSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.CreateServiceAccessor(ServiceIdentifier serviceIdentifier)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService(ServiceIdentifier serviceIdentifier, ServiceProviderEngineScope serviceProviderEngineScope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.GetService(Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
   at Microsoft.Extensions.Logging.OpenTelemetryLoggingExtensions.<>c.<AddOpenTelemetryInternal>b__5_1(IServiceProvider sp) in D:\Source\open-telemetry\codeblanch-opentelemetry-dotnet\src\OpenTelemetry\Logs\ILogger\OpenTelemetryLoggingExtensions.cs:line 246
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite callSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitIEnumerable(IEnumerableCallSite enumerableCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite callSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite callSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite callSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite callSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.CreateServiceAccessor(ServiceIdentifier serviceIdentifier)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService(ServiceIdentifier serviceIdentifier, ServiceProviderEngineScope serviceProviderEngineScope)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService(Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
   at Microsoft.Extensions.Hosting.HostBuilder.<>c__DisplayClass35_0.<PopulateServiceCollection>b__2(IServiceProvider _)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite callSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.CreateServiceAccessor(ServiceIdentifier serviceIdentifier)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService(ServiceIdentifier serviceIdentifier, ServiceProviderEngineScope serviceProviderEngineScope)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService(Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
   at Microsoft.Extensions.Hosting.HostBuilder.ResolveHost(IServiceProvider serviceProvider, DiagnosticListener diagnosticListener)
   at Microsoft.Extensions.Hosting.HostApplicationBuilder.Build()
   at Microsoft.AspNetCore.Builder.WebApplicationBuilder.Build()
   at Program.<Main>$(String[] args) in D:\Source\open-telemetry\codeblanch-opentelemetry-dotnet\examples\AspNetCore\Program.cs:line 164
```
</details>

What if the user isn't using hosting? Well at the moment there isn't a way to get an `IOpenTelemetryBuilder` to use for a call to `UseOtlpExpoter` without hosting 🤣 

#5325 is adding a way to do cross-cutting things without using hosting. If we move forward with that, the current mechanism will crash earlier for that API...

```csharp
     using var openTelemetry = OpenTelemetrySdk.Create(Configure); // NotSupportedException thrown here

     static void Configure(IOpenTelemetryBuilder builder)
     {
            builder
               .ConfigureResource(r => r.AddService("MyService"))
               .UseOtlpExporter();
     }
```